### PR TITLE
Batch Mode

### PR DESCRIPTION
**Pull request by nicholas-leonard**
This PR adds batch mode for SpatialContrastiveNormalization, SpatialDivisiveNormalization, SpatialSubtractiveNormalization and Replicate. Includes unit tests and doc.

Test command:
```lua
th -lnn -e "nn.test{'SpatialDivisiveNormalization_2dkernel','Replicate','SpatialDivisiveNormalization_1dkernel','SpatialSubtractiveNormalization_2dkernel','SpatialSubtractiveNormalization_1dkernel','SpatialContrastiveNormalization'}"
Running 6 tests	
______  ==> Done Completed 44 asserts in 6 tests with 0 errors	
--------------------------------------------------------------------------------	
```

### CODE REVIEW

File: SpatialDivisiveNormalization.lua [hunk 1]
```diff
@@ -82,15 +82,33 @@ function SpatialDivisiveNormalization:__init(nInputPlane, kernel, threshold, thr
...
-      self.coef = self.coef:clone()
+   local dim = input:dim()
+   if self.localstds:dim() ~= self.coef:dim() or (input:size(dim) ~= self.coef:size(dim)) or (input:size(dim-1) ~= self.coef:size(dim-1)) then
+      self.ones = self.ones or input.new()
+      self.__coef = self.__coef or input.new()
+      if dim == 4 then
+         -- batch mode
+         self.ones:resizeAs(input[1]):fill(1)
+         local coef = self.meanestimator:updateOutput(self.ones)
+         self.__coef:resizeAs(coef):copy(coef) -- make contiguous
```
Comment (soumith) [inline SpatialDivisiveNormalization.lua]: the output of meanestimator, is it non-contiguous?

File: SpatialDivisiveNormalization.lua [hunk 2]
```diff
@@ -82,15 +82,33 @@ function SpatialDivisiveNormalization:__init(nInputPlane, kernel, threshold, thr
...
+         self._coef = self._coef or self.coef.new()
+         local coef = self.__coef:view(1,table.unpack(self.__coef:size():totable())):expandAs(self.localstds)
+         self._coef:resizeAs(coef):copy(coef)
```
Comment (soumith) [inline SpatialDivisiveNormalization.lua]: you dont need this resizeAs + copy anymore. the expandAs is good enough (I also should change it in batchnorm).
After the new cuda Apply kernels were landed, which support non-contiguous apply, expanded input which is non-contiguous is perfectly acceptable without introducing a sync point.

File: SpatialSubtractiveNormalization.lua [hunk 3]
```diff
@@ -60,12 +60,30 @@ function SpatialSubtractiveNormalization:__init(nInputPlane, kernel)
...
+      self.ones = self.ones or input.new()
+      self.__coef = self.__coef or input.new()
+      if dim == 4 then
+         -- batch mode
+         self.ones:resizeAs(input[1]):fill(1)
+         local coef = self.meanestimator:updateOutput(self.ones)
+         self.__coef:resizeAs(coef):copy(coef) -- make contiguous
+         self._coef = self._coef or self.coef.new()
+         local size = coef:size():totable()
+         table.insert(size,1,input:size(1))
```
Comment (soumith) [inline SpatialSubtractiveNormalization.lua]: I guess same questions as above.

File: SpatialDivisiveNormalization.lua (hunk 1 above)
Comment (nicholas-leonard) [inline SpatialDivisiveNormalization.lua]: Yes, because of nn.Replicate.

**Comment by nicholas-leonard** [comment]
Made the changes.

**Comment by soumith** [comment]
thanks @nicholas-leonard

**Pull request merged**